appcues/src/main/java/com/appcues/data/model/Experience.kt

```diff
@@ -48,4 +54,34 @@ internal data class Experience(
             }
         }
     }
+
+    // Gets any actions defined on the step group container for the "navigate" trigger. These are the
+    // actions that should be executed before presenting this group's container.
+    fun getNavigationActions(stepContainerIndex: Int): List<ExperienceAction> {
+        val stepGroup = stepContainers.getOrNull(stepContainerIndex) ?: return arrayListOf()
+        return stepGroup.actions[stepGroup.id]?.filter { it.on == NAVIGATE }?.map { it.experienceAction } ?: emptyList()
+    }
+
+    fun areStepsFromDifferentGroup(stepIndexOne: Int, stepIndexTwo: Int): Boolean {
+        return groupLookup[stepIndexOne] != groupLookup[stepIndexTwo]
+    }
+
+    fun getPresentingTrait(flatStepIndex: Int): PresentingTrait {
+        return groupLookup[flatStepIndex]?.let { stepContainers[it].presentingTrait }
+            ?: throw AppcuesTraitException("cannot to find presenting trait")
+    }
+
+    @OptIn(ExperimentalContracts::class)
+    fun isValidStepIndex(flatStepIndex: Int?): Boolean {
+        contract {
+            returns(true) implies (flatStepIndex != null)
+        }
+
+        return flatStepIndex != null && flatStepIndex >= 0 && flatStepIndex < flatSteps.size
+    }
+
+    fun getMetadataSettingTraits(flatStepIndex: Int): List<MetadataSettingTrait> {
```

Comment:
before it this is what we were basically doing to get the step trait, am i missing something here? we can just get the step on the steplist by flatStepIndex i believe

```kotlin
private fun BeginningStep.metadataSettingsTraits(): List<MetadataSettingTrait> {
    return with(experience) {
        // find the container index
        val containerId = groupLookup[flatStepIndex]
        // find the step index in relation to the container
        val stepIndexInContainer = stepIndexLookup[flatStepIndex]
        // if both are valid ids we return Rendering else null
        if (containerId != null && stepIndexInContainer != null) {
            val container = stepContainers[containerId]
            val step = container.steps[stepIndexInContainer]
            // this will throw if metadata fails
            step.metadataSettingTraits
        } else {
            // should be impossible at this point, but will cover with an exception as well
            throw AppcuesTraitException("Invalid step index $flatStepIndex")
        }
    }
}
```

Reply:
Yeah, I think this old code got ported over from somewhere else, and probably not simplified - but I agree, just using the `flatSteps` and the `flatStepIndex` seems like it should work